File: mravi/mravi.py



from math import sqrt


n= int(raw_input())

parent = [0]*(n+1)

perc = [0]*(n+1)

issuper = [0]*(n+1)


for _ in xrange(n-1):
	a,b,x,t = map(int, raw_input().split())
	parent[b]=a
	perc[b] = x
	issuper[b] = t


needs = [0]+map(int, raw_input().split())


best = 0

for i in xrange(1,n+1):
	if needs[i]>=0:
		ans = needs[i]
		while i>1:
			if issuper[i]:
				ans = sqrt(ans)
			ans *= 100.0
			ans /= perc[i]
			i = parent[i]
		best = max(best,ans)


print best

File: mravi/mravi.java

import java.util.*;

public class mravi {
	public static void main(String[] args) {
		Scanner sin = new Scanner(System.in);
		int n = sin.nextInt();
		int [] parent = new int[n+1];
		int [] perc = new int [n+1];
		int [] issuper = new int [n+1];

		for (int i=1;i<n;i++) {
			int a = sin.nextInt();
			int b = sin.nextInt();
			parent[b] = a;
			perc[b] = sin.nextInt();
			issuper[b] = sin.nextInt();
		}

		// int[] needs = new int[n+1];

		double best = 0;

		for (int i =1;i<n+1;i++) {
			double ans = sin.nextDouble();
			if (ans>=0) {
				int j =i;
				while (j>1){
					if (issuper[j]==1) {
						ans = Math.sqrt(ans);
					}
					ans *= (100.0/perc[j]);
					j= parent[j];
				}
				if (ans>best) {
					best=ans;
				}
			}
		}

		System.out.println(best);

	}
}

